async-nats/src/client.rs

```diff
@@ -37,6 +39,63 @@ lazy_static! {
 /// [`Client::publish_with_reply`] or [`Client::publish_with_reply_and_headers`] functions.
 pub struct PublishError(mpsc::error::SendError<Command>);
 
+pub struct Publish {
+    sender: mpsc::Sender<Command>,
+    subject: String,
+    payload: Bytes,
+    headers: Option<HeaderMap>,
+    respond: Option<String>,
+}
+
+impl Publish {
+    pub fn new(sender: mpsc::Sender<Command>, subject: String, payload: Bytes) -> Publish {
+        Publish {
+            sender,
+            subject,
+            payload,
+            headers: None,
+            respond: None,
+        }
+    }
+
+    pub fn headers(mut self, headers: HeaderMap) -> Publish {
+        self.headers = Some(headers);
+        self
+    }
+
+    pub fn reply(mut self, subject: String) -> Publish {
+        self.respond = Some(subject);
+        self
+    }
+}
+
+impl IntoFuture for Publish {
+    type Output = Result<(), PublishError>;
+    type IntoFuture = Pin<Box<dyn Future<Output = Result<(), PublishError>> + Send>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let sender = self.sender.clone();
+        let subject = self.subject;
+        let payload = self.payload;
+        let respond = self.respond;
+        let headers = self.headers;
+
+        Box::pin(async move {
```

Comment:
@caspervonb  I wonder if that additional pinned box has any performance overhead.